### PR TITLE
ghc-with-packages: add an optional exposeGhc parameter (was: Second patchset resulting from the Darwin work)

### PR DESCRIPTION
**Pull request by jwiegley**
This is the second set of minor patches stemming from the Darwin work, even though some of these commits have nothing to do with Darwin.  They were simply things found to be necessary during the course of our work.  Again, I'm submitting them here for review instead of just pushing them directly in case anyone has any objections.  I did not author these commits, but am pinging the people who did in case they would like to offer justification for any of the changes.

Pinging @shlevy @edolstra @peti @joelteon @gridaphobe @copumpkin

### CODE REVIEW

File: pkgs/top-level/release.nix [hunk 1]
```diff
@@ -12,7 +12,7 @@
 { nixpkgs ? { outPath = (import ./all-packages.nix {}).lib.cleanSource ../..; revCount = 1234; shortRev = "abcdef"; }
 , officialRelease ? false
 , # The platforms for which we build Nixpkgs.
-  supportedSystems ? [ "x86_64-linux" "i686-linux" /* "x86_64-darwin" */ ]
+  supportedSystems ? [ "x86_64-linux" "i686-linux" "x86_64-darwin" ]
```
Comment (edolstra) [inline pkgs/top-level/release.nix]: We shouldn't change this until we have sufficient OS X build capacity in the build farm.

**Comment by jwiegley** [comment]
Note that this is the last of the "simple patchsets" that I'm going to be able to push.  The next PR will be a huge one comprising full Yosemite support.  The only thing blocking that PR right now is making sure that it does not break Mavericks support.

File: pkgs/top-level/release.nix (hunk 1 above)
Comment (jwiegley) [inline pkgs/top-level/release.nix]: Ok, reverting this change.  I think we made it for running jobs on our own Hydra at Zalora.

**Comment by wmertens** [comment]
Removing the assert in ccache might cause issues evaluating tarball. Recently master broke for Travis and the fix was to assert away freebsd in the rust compiler.

**Comment by shlevy** [comment]
@wmertens I believe ccache builds on darwin now though.

**Comment by shlevy** [comment]
I'll test the tarball build

**Comment by shlevy** [comment]
I've merged all commits except the last and changed the title accordingly. Waiting on @peti's input.

File: pkgs/development/compilers/ghc/with-packages.nix [hunk 2]
```diff
@@ -44,6 +47,7 @@ buildEnv {
       rm -f $out/bin/$prg
       makeWrapper ${ghc}/bin/$prg $out/bin/$prg         \
         --add-flags '"-B$NIX_GHC_LIBDIR"'               \
+        ${optionalString exposeGHC "--add-flags '-package ghc'"} \
```
Comment (peti) [inline pkgs/development/compilers/ghc/with-packages.nix]: I'm not sure what the addition of this flag is meant to accomplish. What is the use-case for this change?

Comment (shlevy) [inline pkgs/development/compilers/ghc/with-packages.nix]: cc @jwiegley @copumpkin @gridaphobe @joelteon

Comment (jwiegley) [inline pkgs/development/compilers/ghc/with-packages.nix]: I think this originated from @joelteon?

Comment (gridaphobe) [inline pkgs/development/compilers/ghc/with-packages.nix]: I'm the original author. The `ghc` package is hidden by default, which means that you can't just run ghci and start importing GHC modules. It also screws up things like ghc-mod and flycheck. As someone who works a fair bit with the GHC API this is kind of annoying, so I always run `ghc-pkg expose ghc` to un-hide the ghc package.

The point of this flag is to automate that for people like me. I originally wanted to add the `ghc-pkg expose ghc` call to the `postInstall` hook, but ran into a strange permissions issue with `ghc-pkg`, so I took this approach instead.

Comment (peti) [inline pkgs/development/compilers/ghc/with-packages.nix]: In my opinion, `ghcWithPackages` is not the right place for this feature. This addition feels very awkward. Can you not achieve the same thing with a shell alias or a separate wrapper script? Also, adding `ghc` to the list of dependencies in your project's Cabal file would unhide the package as well. Is that kind of local solution not preferable?

Comment (peti) [inline pkgs/development/compilers/ghc/with-packages.nix]: @gridaphobe, if you want to run `ghc-pkg expose ghc` in the `postInstall` hook, then you must first make sure that `$out/lib/ghc-${ghc.version}/package.conf.d` is a writable directory. This should be always the case when you include at least one other Haskell packages other than `ghc` itself in the `ghcWithPackages` environment. The following example works fine for me. I define:

    # ~/.nixpkgs/config.nix

    {
      packageOverrides = super: let self = super.pkgs; in
      {
        ghcDevEnv = (self.haskellPackages.ghcWithPackages (p: with p; [stm])).overrideDerivation (drv: { postBuild = ''
          ${drv.postBuild}
          $out/bin/ghc-pkg expose ghc
        ''; });
      };
    }

and then run:

    $ nix-build -o /tmp/haskell ~/.nix-defexpr --option binary-caches http://hydra.cryp.to -A ghcDevEnv
    these derivations will be built:
      /nix/store/dq4g8w491yj6qblzvscgv2i4fvmdv5a0-haskell-env-ghc-7.8.4.drv
    building path(s) ‘/nix/store/0i532km3pj8b5n87nd9b267zci1mkjw6-haskell-env-ghc-7.8.4’
    building /nix/store/0i532km3pj8b5n87nd9b267zci1mkjw6-haskell-env-ghc-7.8.4
    created 90 symlinks in user environment
    /nix/store/0i532km3pj8b5n87nd9b267zci1mkjw6-haskell-env-ghc-7.8.4

    $ /tmp/haskell/bin/ghci
    GHCi, version 7.8.4: http://www.haskell.org/ghc/  :? for help
    Loading package ghc-prim ... linking ... done.
    Loading package integer-gmp ... linking ... done.
    Loading package base ... linking ... done.
    Prelude> import GHC
    Prelude GHC> :q
    Leaving GHCi.

Works like a charm!

Comment (shlevy) [inline pkgs/development/compilers/ghc/with-packages.nix]: @jwiegley @gridaphobe Any objection to me closing this PR, reverting this change in joelteon/master, and leaving it to @gridaphobe to reopen as a PR unassociated with the darwin changes?

Comment (gridaphobe) [inline pkgs/development/compilers/ghc/with-packages.nix]: Perfectly fine by me. I'll give @peti's `postBuild` hook a try, thanks!

**Pull request closed without merging**